### DISCUSSION TITLE
Fatal Exception: java.lang.UnsatisfiedLinkError

### DISCUSSION THREAD

**Discussion by ComputerDev** [axmolengine/axmol#614] [Q&A] [answered]
Hello,

Since I migrated my project to adxe from cocos2d-x-4.0 I got this error on Android version

I google and can't find a solution.

I saw I maybe have to put -Wl,--hash-style=both but I ignore where to put this :(

Can somebody help me please ?

img1:
![Sans titre2](https://user-images.githubusercontent.com/91702377/161404227-2d8dbcc0-6b97-4ec8-9b92-0d63bf938117.png)

img2:
![Sans titre](https://user-images.githubusercontent.com/91702377/161404126-07aaccf5-c24a-4f42-b1e0-56000abcda6f.png)

Thank you

**Comment by halx99** [comment]
Will take a look later.

**Comment by halx99** [comment]
@ComputerDev  What's your value of `PROP_MIN_SDK_VERSION`?

**Reply by halx99** [reply]
The `libOpenSLES.so` is android system library, and does this issue can reproduce on android emulator 5.1.1?

**Comment by halx99** [comment]
You can try add the link flags as follow:
![image](https://user-images.githubusercontent.com/6977319/161508798-ca81bb8a-8e19-47dc-8365-6ad4ef828bff.png)

And please check the elf hash-style by `aarch64-linux-android-readelf.exe -d your/apk/unpack-path/libopenal.so`, if the flag works, the result will only have `HASH`, no `GNU_HASH`

**Reply by halx99** [reply]
And use the same way to check does your previously libopenal.so only have `GUN_HASH`?

**Comment by ComputerDev** [comment, marked as answer]
Hello,

I have `PROP_MIN_SDK_VERSION=19`

Yes this issue can be reproduce on android emulator 5.1.1 :
![161642137-a9c0f07d-3d40-45f4-aaa6-e8ef59d31e4a](https://user-images.githubusercontent.com/91702377/161643914-1d80ae70-178b-451a-bff6-2994ca1c003d.png)

proj.android\app\build\intermediates\stripped_native_libs\debug\out\lib\x86\libopenal.so :
![Sans titre](https://user-images.githubusercontent.com/91702377/161642541-6e50111a-8ed4-4557-aafe-eae0bae9c32a.png)

with :
```
if (ANDROID)
    set(CMAKE_LINK_DEF_FILE_FLAG "-Wl,--hash-style=both")
    set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -Wl,--hash-style=both")
    set(CMAKE_SHARED_LINKER_FLAGS "${CMAKE_SHARED_LINKER_FLAGS} -Wl,--hash-style=both")
    set(CMAKE_MODULE_LINKER_FLAGS "${CMAKE_MODULE_LINKER_FLAGS} -Wl,--hash-style=both")
endif()
```

I got :
![Sans titre](https://user-images.githubusercontent.com/91702377/161642941-70d04d2b-dc51-4e1f-8c54-0d7639c82548.png)
![Sans titre](https://user-images.githubusercontent.com/91702377/161642878-b1bbf204-101b-4cf0-8bc8-03fc188925c7.png)

But I don't understand, is it not rather libOpenSLES.so that we looking for ?

libOpenSLES.so only have `GNU_HASH` with `aarch64-linux-android-readelf.exe`

Thank you

**Reply by halx99** [reply]
I guess you should release armeabi-v7a .so only for the old device? could you try it?

**Reply by halx99** [reply]
I will try android emulator 5.1.1

**Reply by halx99** [reply]
I have test on android emulator 5.1 (API 22), it seems works (needs change `ANDROID_STL=libc++_shared` in `build.gradle`.
<img width="372" alt="image" src="https://user-images.githubusercontent.com/6977319/161672848-fc2efe06-d305-4afe-be90-2de7b7e219c7.png">

My build arch is `x86`, and my emulator image is `x86`, I will check `PIXEL 2 API 22` as same with your emulator later.

**Reply by halx99** [reply]
Checked, also works on android emulator (`PIXEL 2 API 22 x86_64 image`), and apk build with `x86` only.

<img width="1009" alt="image" src="https://user-images.githubusercontent.com/6977319/161673643-d557a1e6-5a28-4090-80e2-6a5360cb87b8.png">

**Reply by halx99** [reply]
Seems the openal always link with shared c++ library, so I commit the change: https://github.com/adxeproject/adxe/commit/93f549a0ad3ae2dce528dc3c5f7d10978c7f06c6

**Reply by ComputerDev** [reply]
It work! Thank you

**Comment by ComputerDev** [comment]
Hello,

I released my game using your commit without success

```
# set hash style to both for android old device compatible
# see also: https://github.com/adxeproject/adxe/discussions/614
if (ANDROID)
    set(CMAKE_SHARED_LINKER_FLAGS "${CMAKE_SHARED_LINKER_FLAGS} -Wl,--hash-style=both")
endif()
```

![Sans titre](https://user-images.githubusercontent.com/91702377/161652546-1507e74e-01cc-4009-ae5f-9e9d8fa97890.png)

Thank you

**Reply by halx99** [reply]
It's strange the error library is system library `libOpenSLES.so` seems we have no chance rebuild it with `--hash-style=both`
